Describe new spec branch and file layout (#4154)
OAI/OpenAPI-Specification: CONTRIBUTING.md

No changes, ***no matter how trivial***, are ever made to the contents of published specifications.  The only potential changes to those documents are updates to link URLs _if and only if_ the targeted document is moved by a 3rd party.  Other changes to link URLs are not allowed.

### Changing the schemas

Schemas are only changed _after_ the specification is changed.
Changes are made to the YAML versions on the `main` branch.
The JSON versions are generated when published to the [spec site](https://spec.openapis.org/), at which time the `WORK-IN-PROGRESS` URI placeholders are replaced with the publication date.

### Authoritative source of truth

The [spec site](https://spec.openapis.org) is the source of truth.

This changed in 2024, as the markdown files on `main` do not include certain credits and citations.

## Development and publication process

> As of October 2024 (post-OAS 3.0.4 and 3.1.1), the OAS is developed in the `src/oas.md` file on minor release `X.Y-dev` branches that are derived from the baseline `dev` branch.

Reviewer: ```suggestion
As of October 2024 (post-OAS 3.0.4 and 3.1.1), the OAS is developed in the `versions/next.md` file on minor release `X.Y-dev` branches that are derived from the baseline `dev` branch.
```
Why not keep it next to the published versions?

Author: @ralfhandl because one of our problems is that people confuse the published and under-development files.

We want the development to be clearly separated from the publishing.  These are two different functions.

If all of our documentation talks about editing `src/oas.md`, it's much harder to confuse that with `versions/whatever.md`.  Also, the source file isn't a "versions" in the sense of the published files, so putting it in that directory is a category error in my view.

Honestly, my preference would be to not publish through GitHub at all, and _only_ publish directly to `spec.openapis.org`.  That would also allow us to dispense with the `X.Y.Z-rel` branches because the file name wouldn't matter anymore.  But I figured that would be too big of a change (deleting the existing published files off of `main` would probably upset people).

### Branch diagram (3.1.2, 3.2.0, and later)

```mermaid
gitGraph TB:
  commit id:"merge 3.1.1.md to main" tag:"3.1.1"
  branch dev order:1
  commit id:"rename 3.1.1.md to src/oas.md"
  branch v3.1-dev order:2
  commit id:"update version in src/oas.md to 3.1.2"
  checkout dev
  branch v3.2-dev order:5
  commit id:"update version in src/oas.md to 3.2.0"
  commit id:"some 3.2.0 work"
  checkout v3.1-dev
  commit id:"a 3.1.x fix"
  branch v3.1.2-rel order:3
  commit id:"rename src/oas.md to versions/3.1.2.md"
  checkout main
  merge v3.1.2-rel tag:"3.1.2"
  checkout v3.2-dev
  commit id:"more 3.2.0 work"
  checkout v3.1-dev
  commit id:"update version in src/oas.md to 3.1.3"
  commit id:"another 3.1.x fix"
  checkout v3.2-dev
  commit id:"still more 3.2.0 work"
  merge v3.1-dev id:"merge 3.1.x fixes before releasing"
  checkout v3.1-dev
  branch v3.1.3-rel order:4
  commit id:"rename src/oas.md to versions/3.1.3.md"
  checkout v3.2-dev
  branch v3.2.0-rel order:6
  commit id:"rename src/oas.md to versions/3.2.0.md"
  checkout main

>   merge v3.1.3-rel tag:"3.1.3"
>   merge v3.2.0-rel tag:"3.2.0"

Reviewer: This seems to assume that 3.1.3 and 3.2.0 are published at the same time.

Shouldn't it rather be
```suggestion
  merge v3.1.3-rel tag:"3.1.3"
  checkout dev
  merge v3.1-dev id:"update dev with patch release"
  checkout main
  merge v3.2.0-rel tag:"3.2.0"
```

Author: @ralfhandl 

> This seems to assume that 3.1.3 and 3.2.0 are published at the same time.

The point of the example here is to show 3.1.3 and 3.2.0 going out at the same time, as we are likely to do a patch release with each minor release, just like we're doing paired 3.0.4 and 3.1.1 releases.

> Shouldn't it rather be

No, because patch releases don't get merged back to `dev` because they are interleaved across minor release branches.  You can't merge them all consistently.  But we (currently) only ever have one minor release line at a time, so we can merge each minor release down _so it can be the base for the next `X.Y-dev` branch_.  This lets all `X.Y-dev` branches come off of main.  We don't need this for each patch release because they don't get new `-dev` branches, only `-rel` branches which always come off of `dev`.  I'll try to clarify this.

### Branch diagram (3.1.2, 3.2.0, and later)

Initial steps:

* `dev` branches from `main` at the 3.1.1 release commit
* Each `X.Y-dev` branches from `dev`

Upon release:

* Each `X.Y.Z-rel` branches from the corresponding `X.Y-dev`

> * After renaming `src/oad.md`, `X.Y.Z-rel`  merges to `main`

Reviewer: ```suggestion
* After renaming `src/oas.md`, `X.Y.Z-rel`  merges to `main`
```
